docs_nnx/key_concepts.md

```diff
@@ -0,0 +1,320 @@
+
+
+```python
+print(x)
+abs_x = jax.eval_shape(lambda x: x, x)
+print(abs_x)
+```
+
+    [[ 1.0040143  -0.9063372  -0.7481722  -1.1713669 ]
+     [-0.8712328   0.5888381   0.72392994 -1.0255982 ]
+     [ 1.661628   -1.8910251  -1.2889339   0.13360691]
+     [-1.1530392   0.23929629  1.7448074   0.5050189 ]]
+    ShapeDtypeStruct(shape=(4, 4), dtype=float32)
+
+
+It is a good way to dry-run your code and debug a model without any actual compute and memory cost. For example, you can have an overview of the parameters inside this very large model.
+
+
+```python
+class MLP(nnx.Module):
+  def __init__(self, dim, nlayers, rngs):
+    self.blocks = [nnx.Linear(dim, dim, rngs=rngs) for _ in range(nlayers)]
+    self.activation = jax.nn.relu
+    self.nlayers = nlayers
+  def __call__(self, x):
+    for block in self.blocks:
+      x = self.activation(block(x))
+    return x
+
+dim, nlayers = 8190, 64   # Some very big numbers
+@partial(jax.jit, static_argnums=(0, 1))
+def init_state(dim, nlayers):
+  model = MLP(dim, nlayers, nnx.Rngs(0))
+  return nnx.state(model)
+abstract_state = jax.eval_shape(partial(init_state, dim, nlayers))
+print(abstract_state['blocks'][0])
+```
+
+    [38;2;79;201;177mState[0m[38;2;255;213;3m({[0m[38;2;105;105;105m[0m
+      [38;2;156;220;254m'bias'[0m[38;2;212;212;212m: [0m[38;2;79;201;177mVariableState[0m[38;2;255;213;3m([0m[38;2;105;105;105m # 8,190 (32.8 KB)[0m
+        [38;2;156;220;254mtype[0m[38;2;212;212;212m=[0m[38;2;79;201;177mParam[0m,
+        [38;2;156;220;254mvalue[0m[38;2;212;212;212m=[0mShapeDtypeStruct(shape=(8190,), dtype=float32)
+      [38;2;255;213;3m)[0m,
+      [38;2;156;220;254m'kernel'[0m[38;2;212;212;212m: [0m[38;2;79;201;177mVariableState[0m[38;2;255;213;3m([0m[38;2;105;105;105m # 67,076,100 (268.3 MB)[0m
+        [38;2;156;220;254mtype[0m[38;2;212;212;212m=[0m[38;2;79;201;177mParam[0m,
+        [38;2;156;220;254mvalue[0m[38;2;212;212;212m=[0mShapeDtypeStruct(shape=(8190, 8190), dtype=float32)
+      [38;2;255;213;3m)[0m
+    [38;2;255;213;3m})[0m
+
+
+Once you have an abstract pytree for your function input or output, it's easier to describe how you want your data to be sharded. You should use such a pytree with sharding information to instruct your checkpoint loading library to load your arrays distributedly. Our checkpointing guide contains [an example of how to do this](https://flax.readthedocs.io/en/latest/guides/flax_gspmd.html#load-a-sharded-model-from-a-checkpoint).
+
+## Distributed computing
+
+Another big use case for abstract pytrees is to tell JAX machinery how you want each array to be sharded during any point of your computation.
```

Comment:
Would be nice to introduce concept by itself shortly and only mention that `jax.lax.with_sharding_constrain` accepts pytrees as a comment (hopefully the reader gets the idea by now).